doc/languages-frameworks/rust.md

```diff
@@ -57,6 +57,154 @@ checksum can be then take from the failed build.
 [nixcrates](https://github.com/fractalide/nixcrates).
 
+## Compiling Rust crates using Nix instead of Cargo
+
+When run, `cargo build` produces a file called `Cargo.lock`,
+containing pinned versions of all dependencies. Nixpkgs contains a
+tool called `carnix` (`nix-env -iA nixos.carnix`), which can be used
+to turn a `Cargo.lock` into a Nix expression.
+
+That Nix expression calls `rustc` directly (hence bypassing Cargo),
+and can be used to compile a crate and all its dependencies. Here is
+an example for a minimal `hello` crate:
+
+
+    $ cargo new hello
+    $ cd hello
+    $ cargo build
+     Compiling hello v0.1.0 (file:///tmp/hello)
+      Finished dev [unoptimized + debuginfo] target(s) in 0.20 secs
+    $ carnix -o hello.nix --src ./. Cargo.lock --standalone
+    $ nix-build hello.nix
+
+Now, the file produced by the call to `carnix`, called `hello.nix`, looks like:
+
+```
+with import <nixpkgs> {};
+let release = true;
+    verbose = true;
+    hello_0_1_0_ = { dependencies?[], buildDependencies?[], features?[] }: buildRustCrate {
+      crateName = "hello";
+      version = "0.1.0";
+      src = ./.;
+      inherit dependencies buildDependencies features release verbose;
+    };
+
+in
+rec {
+  hello_0_1_0 = hello_0_1_0_ {};
+}
+```
+
+In particular, note that the argument given as `--src` is copied
+verbatim to the source.  If we look at a more complicated
+dependencies, for instance by adding a single line `libc="*"` to our
+`Cargo.toml`, we first need to run `cargo build` to update the
+`Cargo.lock`. Then, `carnix` needs to be run again, and produces the
+following nix file:
+
+```
+with import <nixpkgs> {};
+let release = true;
+    verbose = true;
+    hello_0_1_0_ = { dependencies?[], buildDependencies?[], features?[] }: buildRustCrate {
+      crateName = "hello";
+      version = "0.1.0";
+      src = ./.;
+      inherit dependencies buildDependencies features release verbose;
+    };
+    libc_0_2_33_ = { dependencies?[], buildDependencies?[], features?[] }: buildRustCrate {
+      crateName = "libc";
+      version = "0.2.33";
+      sha256 = "1l7synziccnvarsq2kk22vps720ih6chmn016bhr2bq54hblbnl1";
+      inherit dependencies buildDependencies features release verbose;
+    };
+
+in
+rec {
+  hello_0_1_0 = hello_0_1_0_ {
+    dependencies = [ libc_0_2_33 ];
+  };
+  libc_0_2_33 = libc_0_2_33_ {
+    features = [ "use_std" ];
+  };
+}
+```
+
+Here, the `libc` crate has no `src` attribute, so `buildRustCrate`
+will fetch it from [crates.io](https://crates.io). A `sha256`
+attribute is still needed for Nix purity.
+
+Some crates require external libraries. For crates from
+[crates.io](https://crates.io), such libraries can be specified in
+`<nixpkgs/pkgs/build-support/rust/defaultCrateOverrides.nix>`.
+
+Starting from that file, one can add more overrides, to add features
+or build inputs, as follows:
+
+```
+let
+  defaultOverrides = callPackage <nixpkgs/pkgs/build-support/rust/defaultCrateOverrides.nix> {};
```

Comment:
The ruby infrastructure exposes this file as a package attribute. Can you do the same, and use that example? That way, `nixpkgs` can be overridden without having to actually replace the entry in the NIX_PATH.

Reply:
I'm not sure I understood your comment correctly, is my latest commit a fix? A "package attribute" would be an attribute of `pkgs`, or of `carnix` itself?

Reply:
This is already the case: https://github.com/NixOS/nixpkgs/pull/31150/files#diff-036410e9211b4336186fc613f7200b12R6224
It is allready a package in the all-packages namespace.
Just the documentation should be updated to use it instead of using `callPackage`.